### PR TITLE
More binary http message supports in SpringEncoder

### PR DESCRIPTION
**Pull request by ScienJus**
Hi guys, I'm trying to use Feign with Protobuf, but I found Feign is incompatible with Protobuf. So I write `ProtobufHttpMessageConverter`.

```
public class ProtobufHttpMessageConverter extends AbstractHttpMessageConverter<Message> {
    
    private static final ConcurrentHashMap<Class<?>, Method> METHOD_CACHE = new ConcurrentHashMap<Class<?>, Method>();

    public static final String X_PROTOBUF_SCHEMA_HEADER = "X-Protobuf-Schema";

    public static final String X_PROTOBUF_MESSAGE_HEADER = "X-Protobuf-Message";

    private final ExtensionRegistry extensionRegistry = ExtensionRegistry.newInstance();

    public ProtobufHttpMessageConverter() {
        super(new MediaType("application", "x-protobuf"));
    }

    @Override
    protected boolean supports(Class<?> clazz) {
        return Message.class.isAssignableFrom(clazz);
    }

    @Override
    protected Message readInternal(Class<? extends Message> clazz, HttpInputMessage inputMessage) throws IOException, HttpMessageNotReadableException {
        try {
            Message.Builder builder = getMessageBuilder(clazz);
            builder.mergeFrom(inputMessage.getBody(), this.extensionRegistry);
            return builder.build();
        } catch (Exception e) {
            throw new HttpMessageNotReadableException("Could not read Protobuf message: " + e.getMessage(), e);
        }
    }

    @Override
    protected void writeInternal(Message message, HttpOutputMessage outputMessage) throws IOException, HttpMessageNotWritableException {
        setProtoHeader(outputMessage, message);
        FileCopyUtils.copy(message.toByteArray(), outputMessage.getBody());
    }

    /**
     * Set the "X-Protobuf-*" HTTP headers when responding with a message of
     * content type "application/x-protobuf"
     * <p><b>Note:</b> <code>outputMessage.getBody()</code> should not have been called
     * before because it writes HTTP headers (making them read only).</p>
     */
    private void setProtoHeader(HttpOutputMessage response, Message message) {
        response.getHeaders().set(X_PROTOBUF_SCHEMA_HEADER, message.getDescriptorForType().getFile().getName());
        response.getHeaders().set(X_PROTOBUF_MESSAGE_HEADER, message.getDescriptorForType().getFullName());
    }

    /**
     * Create a new {@code Message.Builder} instance for the given class.
     * <p>This method uses a ConcurrentHashMap for caching method lookups.
     */
    private static Message.Builder getMessageBuilder(Class<? extends Message> clazz) throws Exception {
        Method method = METHOD_CACHE.get(clazz);
        if (method == null) {
            synchronized (clazz) {
                method = METHOD_CACHE.get(clazz);
                if (method == null) {
                    method = clazz.getMethod("newBuilder");
                    METHOD_CACHE.put(clazz, method);
                }
            }
        }
        return (Message.Builder) method.invoke(clazz);
    }
}
```

But I have encountered a problem while developing this feature. The HTTP request-body bytes is different from the original Protobuf object serialized bytes.

After troubleshooting I found that `SpringEncoder` would use `UTF-8` charset when converter is not `ByteArrayHttpMessageConverter`:

```
// do not use charset for binary data
if (messageConverter instanceof ByteArrayHttpMessageConverter) {
    request.body(outputMessage.getOutputStream().toByteArray(), null);
} else {
    request.body(outputMessage.getOutputStream().toByteArray(), Charset.forName("UTF-8"));
}
```

The body will be encoded as StringEntity by `ApacheHttpClient` when Request contains charset, which has changed the original body.

```
if (request.body() != null) {
    HttpEntity entity = null;
    if (request.charset() != null) {
        ContentType contentType = getContentType(request);
        String content = new String(request.body(), request.charset());
        entity = new StringEntity(content, contentType);
    } else {
        entity = new ByteArrayEntity(request.body());
    }

    requestBuilder.setEntity(entity);
}
```

In conclusion, I wrote an interface `BinaryHttpMessageConvert` which could indicate that there is no need to set charset since the `ByteArrayHttpMessageConverter` does not support generic type.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@ScienJus Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=2446)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=2446) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@ScienJus Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=2446)!

**Comment by ScienJus** [comment]
@spencergibb Thanks for your reply. My `ProtobufHttpMessageConverter` is a simplified version of spring's, so I think it will have the same problem in feign. I will try to test it.

**Comment by ScienJus** [comment]
@spencergibb 

I have created a few tests that demonstrate that Spring's `ProtobufHttpMessageConverter` fires the same bug in SpringEncoder, The same reason is encoded and charset.

The test cases in this [repository](https://github.com/ScienJus/spring-cloud-feign-protobuf). Can you take a look?

At the same time, I realized that this fix(add a new interface) can not solve the problem. So do you have any good solution?

The important thing is that we need to fix the wrong set of encodings in the `SpringEncoder`.

Thanks!

**Comment by gdong42** [comment]
When will this be available? Having similar problem here.
